torch/optim/adagrad.py

```diff
@@ -40,6 +40,10 @@ def __init__(self, params, lr=1e-2, lr_decay=0, weight_decay=0, initial_accumula
 
         for group in self.param_groups:
             for p in group['params']:
+                if isinstance(p, torch.nn.parameter.UninitializedParameter):
+                    raise ValueError(
+                        '`torch.optim.AdaGrad` does not support torch.nn.parameter.UninitializedParameter'
+                        'Make sure to run the dummy forward before creating the optimizer.')
```

Comment:
Is it correct that we only need to check this in `Adagrad` and not other optimizers? For user-created optimizers in the wild, currently is there any mechanism to hard-fail and mention the need of dummy forward pass?

In general I am thinking whether we should put more of these checks into the `UninitializedParameter` class methods rather than in the use sites (probably except for a few cases where we have to catch these errors early)

Reply:
> In general I am thinking whether we should put more of these checks into the UninitializedParameter class methods rather than in the use sites

I think that from the user experience side, having a clear error telling them what to do when they try to use it is quite important and is a good reason to put these errors at the use site.

> Is it correct that we only need to check this in Adagrad and not other optimizers?

Yes. There are tests for the other optimizers to make sure they work fine.

> For user-created optimizers in the wild, currently is there any mechanism to hard-fail and mention the need of dummy forward pass?

Unfortunately no. And because they could be creating states in any way they want, it is hard to say how to catch these. Hopefully, people following the example from our basic optimizers should be lazily initializing buffers and it will work fine.

Reply:
I wonder whether we should override `__torch_function__` for UninitializedParameter class. That should give good error messages in attempting to use this op in any of the torch operations. E.g. in `torch.full_like` below. Have we considered that?

Reply:
I jus tried this but it was pretty tricky to filter accessing to `param.grad_fn` or `param.data.device` without giving errors...

Reply:
pushed :)

Reply:
@emcastillo curious with the `__torch_function__ ` change, do we still need to keep the `UninitializedParameter` check in other places in this PR? (happy with either decision, just wanted to make sure it's an intentional one :)

Reply:
I think that a hard-error in dataparallel, or the weight norms, is more user-friendly than just failing in a random torch function that the user is not aware of, so I think both checks can co-exist.